engine/api/src/main/scala/pl/touk/nussknacker/engine/api/dict/DictDefinition.scala

```diff
@@ -0,0 +1,44 @@
+package pl.touk.nussknacker.engine.api.dict
+
+import pl.touk.nussknacker.engine.api.typed.typing.{SingleTypingResult, Typed, TypedClass}
+
+/**
+ * It defines dictionary which will contain values of type `valueType`. Value will be created
+ * based on String `key`. `key` is presented as a `label` in UI.
+ *
+ * All definitions should be registered in `ProcessConfigCreator.expressionConfig().dictionaries`
+ * After registration, definitions will be available for `DictRegistry` and other services using it (like `DictQueryService`)
+ */
+trait DictDefinition extends Serializable {
+
+  def valueType: SingleTypingResult
+
+  // It should return value in declared type.
+  def value(key: String): Any
+
+}
+
+/**
+ * If this instance wil be used in global variables, will be typed to TypedDict.
+ */
+case class DictInstance(dictId: String, definition: DictDefinition) {
+
+  def valueType: SingleTypingResult = definition.valueType
+
+  def value(key: String): Any = definition.value(key)
+
+}
+
+/**
+ * It is helper mixin when value is exact as key (without additional transformation)
+ */
+trait ReturningKeyWithoutTransformation { self: DictDefinition =>
```

Comment:
oook... I kind of understand what you want to achieve now...
I think we should maybe add some helpers or checks (in ProcessConfigCreation) to make sure that dictIds are the same as global var names? Because now it can be *very* confusing I think ;)

Reply:
There are three things:
1. duplication of dictId between DictInstance and SimpleDictDefinition -> I've fixed it by adding dictId to valueType as a parameter
2. duplication of dictId between DictInstance and `ExpressionConfig.dictionaries` (definitions are in map like sources etc and we have there categories so it will be hard to change that). I can add TODO that after category refactor we can move dictId to `DictDefinition` and then this duplication will disappear.
3. DictInstance's dictId vs global variable name - IMO it's not a duplication, because developer can organize dict instance's in many ways e.g. he or she can put it to nested variable "dicts" as a properties.

Regards to 3. how to assert 2.? One idea that I have is to add factory method to `DictInstance` which will get `Map[String, WithCategories[DictDefinition]]` and `dictId` and create `DictInstance`. It will be ugly IMO :-( Any thoughts?

Reply:
let's add TODO...